### PR TITLE
Added discrete variable support to Metropolis

### PR DESCRIPTION
**Pull request by fonnesbeck**
`Metropolis` now supports discrete as well as continuous variables. Added a `BinaryMetropolis` for use with Bernoulli random variables. Updated `latent_occupancy` example to use `BinaryMetropolis`.

### CODE REVIEW

File: pymc/step_methods/metropolis.py [hunk 1]
```diff
@@ -119,4 +136,42 @@ def tune(scale, acc_rate):
...
+        q = copy(q0)
+        # Locations where switches occur, according to p_jump
+        switch_locs = where(rand_array<p_jump)
+        q[switch_locs] = True - q[switch_locs]
+
+        q_new = metrop_select(logp(q) - logp(q0), q, q0)
+
+        return q_new
+
+def subst(a, val, index):
```
Comment (jsalvatier) [inline pymc/step_methods/metropolis.py]: This doesn't look like it's called anywhere.

File: pymc/step_methods/metropolis.py [hunk 2]
```diff
@@ -119,4 +136,42 @@ def tune(scale, acc_rate):
...
+        self.steps_until_tune = tune_interval
+        self.accepted = 0
+
+        if not all([v.dtype.startswith('int') for v in vars]):
+            raise ValueError('All variables must be Bernoulli for BinaryMetropolis')
+
+        super(BinaryMetropolis,self).__init__(vars, [model.logpc])
+
+    def astep(self, q0, logp):
+        """docstring for astep"""
```
Comment (jsalvatier) [inline pymc/step_methods/metropolis.py]: Probably better not to have placeholder docstrings

File: pymc/step_methods/metropolis.py [hunk 3]
```diff
@@ -59,6 +64,15 @@ def __init__(self, vars, S=None, proposal_dist=normal_proposal, scaling=1.,
         self.tune_interval = tune_interval
         self.steps_until_tune = tune_interval
         self.accepted = 0
+
+        # Determine type of variables
+        if all([v.dtype=='int64' for v in vars]):
```
Comment (jsalvatier) [inline pymc/step_methods/metropolis.py]: I think you should use `v.dtype in continuous_types` and `v.dtype in discrete_types` from https://github.com/pymc-devs/pymc/blob/pymc3/pymc/types.py

File: pymc/step_methods/metropolis.py [hunk 4]
```diff
@@ -119,4 +136,42 @@ def tune(scale, acc_rate):
...
+
+        model = modelcontext(model)
+
+        self.scaling = scaling
+        self.tune = tune
+        self.tune_interval = tune_interval
+        self.steps_until_tune = tune_interval
+        self.accepted = 0
+
+        if not all([v.dtype.startswith('int') for v in vars]):
```
Comment (jsalvatier) [inline pymc/step_methods/metropolis.py]: We should either check for int8 (the type used for binary in Theano) or for `var.dtype in discrete_types`

**Comment by jsalvatier** [comment]
Looks good! It'll be great to have more support for discrete variables. Looks like you have to rebase before we merge.

**Comment by fonnesbeck** [comment]
Thanks for catching those. Should be fixed in 0aba1e9f608e594baec854b61283238b9a2ab116.

**Comment by jsalvatier** [comment]
Looks good to me, I'll merge when you rebase (or you can).

**Pull request merged**